# **Exercise 1**

In [ ]:
hexa ='A''B''C''D''E''F''a''b''c''d''e''f'
deci ='2''3''4''5''6''7''8''9'
bins = '0''1'
input1 = input()

def IsBins(inpt):
  for char in inpt:
    if char in bins:
     return True

def Isdecis(inpt):
  for char in inpt:
    if char in deci:
      return True

def IsHexs(inpt):
    for char in inpt:
      if char in hexa:
        return True

if IsHexs(input1):
  input_3 = int(input1,16)
  print("Your input is Hexadecimal.")
elif Isdecis(input1):
  input_3 = int(input1) 
  print("Your input is Decimal.")
elif IsBins(input1):
  input_3 = int(input1,2)
  print("Your input is Binary.")

input2= input("Convert it to: \n 1)Binary \n 2)Decimal \n 3)Hexadecimal")

if input2 == '1':
  print(bin(input_3))
elif input2 == '2':
  print(int(input_3))
elif input2 == '3':
  print(hex(input_3))


2345
Your input is Decimal.
Convert it to: 
 1)Binary 
 2)Decimal 
 3)Hexadecimal1
0b100100101001


# **Exercise 2**

In [ ]:
def Convert(strs):
  power = -1
  mantissa = 0

  for i in strs:
    mantissa += (int(i) * pow(2, power))
    power -= 1    
  return (mantissa + 1)


ieee32 = input("Write the 32 bit binary string: ")
sign = int(ieee32[0])
bias = int(ieee32[1 : 9], 2)
unbias = bias - 127

mantstr = ieee32[9 : ]
mantint = Convert(mantstr)
realnumber = pow(-1, sign) * mantint * pow(2, unbias)

print("Float representation",realnumber)

Write the 32 bit binary string: 110000101011000000000000
Float representation -88.0


# **Exercise 3**

In [ ]:
n=1023
underflow = 1
overflow = 1
for i in range (n):
  underflow = underflow/2
  overflow = overflow*2
  print("|%2d"%i, "|","\t\t","|%2.5e"%underflow, "|", "\t\t", "|%2.5e"%overflow, "|")
# The underflow and overflow limits on my computer is 2**1023

| 0 | 		 |5.00000e-01 | 		 |2.00000e+00 |
| 1 | 		 |2.50000e-01 | 		 |4.00000e+00 |
| 2 | 		 |1.25000e-01 | 		 |8.00000e+00 |
| 3 | 		 |6.25000e-02 | 		 |1.60000e+01 |
| 4 | 		 |3.12500e-02 | 		 |3.20000e+01 |
| 5 | 		 |1.56250e-02 | 		 |6.40000e+01 |
| 6 | 		 |7.81250e-03 | 		 |1.28000e+02 |
| 7 | 		 |3.90625e-03 | 		 |2.56000e+02 |
| 8 | 		 |1.95312e-03 | 		 |5.12000e+02 |
| 9 | 		 |9.76562e-04 | 		 |1.02400e+03 |
|10 | 		 |4.88281e-04 | 		 |2.04800e+03 |
|11 | 		 |2.44141e-04 | 		 |4.09600e+03 |
|12 | 		 |1.22070e-04 | 		 |8.19200e+03 |
|13 | 		 |6.10352e-05 | 		 |1.63840e+04 |
|14 | 		 |3.05176e-05 | 		 |3.27680e+04 |
|15 | 		 |1.52588e-05 | 		 |6.55360e+04 |
|16 | 		 |7.62939e-06 | 		 |1.31072e+05 |
|17 | 		 |3.81470e-06 | 		 |2.62144e+05 |
|18 | 		 |1.90735e-06 | 		 |5.24288e+05 |
|19 | 		 |9.53674e-07 | 		 |1.04858e+06 |
|20 | 		 |4.76837e-07 | 		 |2.09715e+06 |
|21 | 		 |2.38419e-07 | 		 |4.19430e+06 |
|22 | 		 |1.19209e-07 | 		 |8.38861e+06 |
|23 | 		 |5.96046e-08 | 		 |1.6777

# **Exercise 4**

In [ ]:
n = 4
n2 = 23
for i in range(n2):
  n += (0.1)**i
  print("|%2d"%i, "|","\t\t","|",n)

| 0 | 		 | 5.0
| 1 | 		 | 5.1
| 2 | 		 | 5.109999999999999
| 3 | 		 | 5.111
| 4 | 		 | 5.1110999999999995
| 5 | 		 | 5.111109999999999
| 6 | 		 | 5.111110999999999
| 7 | 		 | 5.1111111
| 8 | 		 | 5.1111111099999995
| 9 | 		 | 5.111111111
|10 | 		 | 5.1111111111
|11 | 		 | 5.11111111111
|12 | 		 | 5.111111111111
|13 | 		 | 5.1111111111111
|14 | 		 | 5.11111111111111
|15 | 		 | 5.111111111111111
|16 | 		 | 5.111111111111111
|17 | 		 | 5.111111111111111
|18 | 		 | 5.111111111111111
|19 | 		 | 5.111111111111111
|20 | 		 | 5.111111111111111
|21 | 		 | 5.111111111111111
|22 | 		 | 5.111111111111111


The machine precision point for floating numbers is 16.

# **Exercise 5**

a)

In [ ]:
import numpy as np

a0,b0,c0 = input().split(' ')
a = float(a0)
b = float(b0)
c = float(c0)

def rootcalculator(a1,b1,c1):
  dRoot = np.sqrt(b1**2-4*a1*c1)
  x1 = ((-b1) + dRoot)/(2*a1)
  x2 = ((-b1) - dRoot)/(2*a1)
  print('x1 is : ', x1, ' \n and x2 is : ', x2)

rootcalculator(a,b,c)

0.001 1000 0.001
x1 is :  -9.999894245993346e-07  
 and x2 is :  -999999.999999


**b)**

In [ ]:

a0,b0,c0 = input().split(' ')
a = float(a0)
b = float(b0)
c = float(c0)

def rootcalculator(a1,b1,c1):
  dRoot = np.sqrt(b1**2-4*a1*c1)
  x1 = (((-b1) + dRoot) * ((-b1)-dRoot)) / ((2*a1)*((-b1) - dRoot))
  x2 = (((-b1) - dRoot) * ((-b1)+dRoot)) / ((2*a1)*((-b1) + dRoot))
  print('x1 is : ', x1, ' \n and x2 is : ', x2)

rootcalculator(a,b,c)

0.001 1000 0.001
x1 is :  -9.999894245993346e-07  
 and x2 is :  -999999.9999990001


The first $x_{1} =  -9.999894245993346e-07 $  and the second one is $x_{1} = -9.999894245993346e-07 $
The first one is $x_{2} = -999999.999999$ and the second one is $x_{2} =  -999999.9999990001 $
It is because the computers store numbers in binary format.

**c)**

In [ ]:

a0,b0,c0 = input().split(' ')
a = float(a0)
b = float(b0)
c = float(c0)

def rootcalculator(a1,b1,c1):
  dRoot = (np.sqrt(b1**2-4*a1*c1))
  x1 = (((-b1) + dRoot)/(2*a1))
  x2 = (((-b1) - dRoot)/(2*a1))
  print('x1 is : ', x1.hex(), ' \n and x2 is : ', x2.hex())

rootcalculator(a,b,c)

0.001 1000 0.001
x1 is :  -0x1.0c6ec00000000p-20  
 and x2 is :  -0x1.e847fffffde72p+19


In [ ]:
a0,b0,c0 = input().split(' ')
a = float(a0)
b = float(b0)
c = float(c0)

def rootcalculator(a1,b1,c1):
  dRoot = np.sqrt(b1**2-4*a1*c1)
  x1 = (((-b1) + dRoot) * ((-b1)-dRoot)) / ((2*a1)*((-b1) - dRoot))
  x2 = (((-b1) - dRoot) * ((-b1)+dRoot)) / ((2*a1)*((-b1) + dRoot))
  print('x1 is : ', x1.hex(), ' \n and x2 is : ', x2.hex())

rootcalculator(a,b,c)

0.001 1000 0.001
x1 is :  -0x1.0c6ec00000000p-20  
 and x2 is :  -0x1.e847fffffde73p+19


The first and the second answer is same  is $x_{1} = -0x1.0c6ec00000000p-20 $ and $x_{2} = -0x1.e847fffffde73p+19 $



# **Exercise 6**


a)

In [ ]:
def f(x):
 return x*(x-1) 

def Derivativef(x,h):
  deriv = (f(x+h)-f(x))/h
  return deriv
  
print(Derivativef(1,10**(-2)))

1.010000000000001


In [ ]:
def derivanalytical(x):
  return 2*x - 1
print(derivanalytical(1))

1


THe answers are different because in first approach we are using limit to gain an approxiamted answer.

b)

In [ ]:
print("\n For 10**-4 : ",Derivativef(1,10**(-4)),"\n   difference is", 1-Derivativef(1,10**(-4)) )
print("\n For 10**-6 : ",Derivativef(1,10**(-6)),"\n   difference is", 1-Derivativef(1,10**(-6)) )
print("\n For 10**-8 : ",Derivativef(1,10**(-8)),"\n   difference is", 1-Derivativef(1,10**(-8)) )
print("\n For 10**-10 : ",Derivativef(1,10**(-10)),"\n   difference is", 1-Derivativef(1,10**(-10)) )
print("\n For 10**-12 : ",Derivativef(1,10**(-12)),"\n   difference is", 1-Derivativef(1,10**(-12)) )
print("\n For 10**-14 : ",Derivativef(1,10**(-14)),"\n   difference is", 1-Derivativef(1,10**(-14)) )



 For 10**-4 :  1.0000999999998899 
   difference is -9.999999988985486e-05

 For 10**-6 :  1.0000009999177333 
   difference is -9.99917733279787e-07

 For 10**-8 :  1.0000000039225287 
   difference is -3.922528746258536e-09

 For 10**-10 :  1.000000082840371 
   difference is -8.284037100736441e-08

 For 10**-12 :  1.0000889005833413 
   difference is -8.890058334132256e-05

 For 10**-14 :  0.9992007221626509 
   difference is 0.0007992778373491216


# **Ecercise 7**

In [ ]:
import numpy as np

def f(x):
  return np.sqrt(1-x**2)
N=100
h=2/N          
I=0            
x=-1           
for i in range(N): 
    I+= h*f(x)
    x+= h
print("For N=100:\n",I)
print("True value:\n", np.pi/2)

Calculated value for N=100:
 1.5691342555492493
True value:
 1.5707963267948966
